lib/probes/mongodb.js

```diff
@@ -1,347 +1,340 @@
-    data.Group_Initial = JSON.stringify(cmd.group.initial)
+  group: (cmd) => {
+    const data = {
+      QueryOp: 'group',
+      Group_Condition: prepQueryString(cmd.group.cond),
+      Group_Initial: prepQueryString(cmd.group.initial),
+      Group_Key: prepQueryString(cmd.group.key)
+    }
+
     if (typeof cmd.group.$reduce === 'function') {
       data.Group_Reduce = cmd.group.$reduce.toString()
     } else if (typeof cmd.group.$reduce === 'object') {
       data.Group_Reduce = cmd.group.$reduce.code.toString()
     } else {
       data.Group_Reduce = cmd.group.$reduce.toString()
     }
-    data.Group_Key = JSON.stringify(cmd.group.key)
-  }],
-  [ifDef('mapReduce'), function (data, cmd) {
-    data.QueryOp = 'map_reduce'
-    data.Map_Function = cmd.map
-    data.Reduce_Function = cmd.reduce
+
+    return data
+  },
+  mapReduce: (cmd) => {
+    const data = {
+      QueryOp: 'map_reduce',
+      Map_Function: cmd.map,
+      Reduce_Function: cmd.reduce
+    }
+
+    if (cmd.finalize) {
+      data.Finalize_Function = cmd.finalize
+    }
+
+    return data
+  },
+  map: (cmd) => { // alias for mapReduce used in earlier versions of 3.x
+    const data = {
+      QueryOp: 'map_reduce',
+      Map_Function: cmd.map,
+      Reduce_Function: cmd.reduce
+    }
+
     if (cmd.finalize) {
       data.Finalize_Function = cmd.finalize
     }
-  }],
-  [ifDef('aggregate'), function (data, cmd) {
-    data.QueryOp = 'aggregate'
-    data.Pipeline = JSON.stringify(cmd.pipeline)
-  }]
-]
-
-function makeData (ctx, ns, cmd) {
-  const data = makeBaseData(ctx, ns)
-  for (let i = 0; i < dataMakers.length; i++) {
-    const [test, make] = dataMakers[i]
-    if (test(cmd)) {
-      make(data, cmd)
-      return data
+
+    return data
+  },
+  aggregate: (cmd) => {
+    return {
+      QueryOp: 'aggregate',
+      Pipeline: prepQueryString(cmd.pipeline)
+    }
+  },
+
+  // atomic commands in 3.x
+  insert: (opsOrCmd) => {
+    return {
+      QueryOp: 'insert',
+      Insert_Document: prepQueryString((opsOrCmd))
+    }
+  },
+  update: (opsOrCmd) => {
+    const obj = Array.isArray(opsOrCmd) ? opsOrCmd : opsOrCmd.updates
+    return {
+      QueryOp: 'update',
+      Query: prepQueryString(obj.map(v => v.q)),
+      Update_Document: prepQueryString(obj.map(v => v.u))
+    }
+  },
+  remove: (opsOrCmd) => {
+    const obj = Array.isArray(opsOrCmd) ? opsOrCmd : opsOrCmd.updates
+    return {
+      QueryOp: 'remove',
+      Query: prepQueryString(obj.map(v => v.q))
+    }
+  },
+
+  // deprecated but may still be used
+  delete: (opsOrCmd) => {
+    const obj = Array.isArray(opsOrCmd) ? opsOrCmd : opsOrCmd.deletes
+    return {
+      QueryOp: 'remove',
+      Query: prepQueryString(obj.map(v => v.q))
     }
   }
+}
 
-  data.QueryOp = 'command'
-  data.Command = JSON.stringify(cmd)
+module.exports = function (mongodb, options) {
+  // v3.3.0 is the first version of mongodb that doesn't use mongodb-core.
```

Comment:
_Basic_ questions from me. Not sure where this is in the code and/or compatibility chart.

The agent supports Mongo `2.6` and `3.0`, `3.1`, `3.2` as well, which I think do use `mongodb-core` is that right? Generally, what happens in the agent if customer is using one of those older versions with this PR update? e.g. Are all commands and values sanitized the same way as `>=3.3.0`? How come `requires('mongodb-core')` is no longer needed?

Reply:
mongodb the *database* versions 2.6 to 5
mongodb the *driver* https://www.npmjs.com/package/mongodb 3.3.0 and above.

Customer who has older version of *the database* should *upgrade the driver* using `npm install mongodob@legacy` (which will give them driver 3.7.3). Then from instrumentation point of view the 'mongodb-core' package is dropped. In short - we do not support outdated legacy configurations.

Reply:
Hah, key piece I was missing. Thank you!